### PR TITLE
Improve VR telemetry integration with developer GA tracking

### LINKED ISSUE #199: Fix compatibility problems with GA
Ease GA integration by allowing to use the name `ga` for the tracker function returned by `MozillaResearch.telemetry.ga.create` method since that's the popular name used in the Google Analytics guides. Also, allow to use routed messages.

### PR DESCRIPTION
**Pull request by delapuente**
Fix #199

### CODE REVIEW

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 1]
```diff
@@ -54,10 +55,11 @@ telemetry.start = function (config) {
 setupAnalytics();
 
 function setupAnalytics() {
-  window.ga = window.ga || function () {
-    (window.ga.q = (window.ga.q || [])).push(arguments)
+  window.GoogleAnalyticsObject = '_ga';
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: FWIW, other Mozilla properties block on DNT: https://github.com/MozillaReality/vr-mozilla-org/blob/4dbef3437/static/js/ga.js#L5-L6

should we add a comment or block all GA behind DNT (I know not everyone agrees that GA should respect DNT, which it doesn't out of the box)

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 2]
```diff
@@ -54,10 +55,11 @@ telemetry.start = function (config) {
...
+  window._ga = window._ga || function () {
+    (window._ga.q = (window._ga.q || [])).push(arguments)
   };
-  window.ga.l = +(new Date());
+  window._ga.l = +(new Date());
 
   if (navigator.doNotTrack === '1') {
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: per my comment above, can we/should we move this above?

**Comment by delapuente** [comment]
@cvan I made the code more restrictive when dealing with Do-Not-Track. My criteria is documented in the library and it's as follows:

```js
/*
 * The main requirements when dealing with Do-Not-Track (DNT) are:
 * 
 *  1. Not to break the API.
 *  2. Prevent explicit collection, storage and sending data if DNT is enabled.
 *  3. Only sessions where DNT is disabled for its entire duration are expected to send
 *     telemetry data.*
 * 
 * Because some collection can happen implicitly when loading 3rd party libraries. It is
 * also nice to have preventing loading external libraries if DNT is enabled but it must be
 * done in a way not conflicting with the requirement 1.
 * 
 * (*) In ther words: it is not required to resume collecting telemetry if DNT has been
 * disabled after the session started.
 */
```

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 3]
```diff
@@ -134,4 +156,8 @@ function injectScript (src, callback) {
   return script;
 }
 
+function doNotTrack() {
+  return navigator.doNotTrack === '1';
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I just learned from [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/doNotTrack#Browser_compatibility) that IE 9/10 expose as `navigator.msDoNotTrack` and IE 11/Edge expose as `window.doNotTrack`. lol, I don't know why they went from a vendor prefix to a proprietary property on `window` instead of `navigator`.

so you can just change to this:

```js
  // `navigator.doNotTrack` for spec-compliant browsers.
  // `navigator.msDoNotTrack` for IE 9/10.
  // `window.msDoNotTrack` for IE 11/Edge.
  return navigator.doNotTrack === '1' || navigator.msDoNotTrack === '1' || window.doNotTrack === '1';
```

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 4]
```diff
@@ -13,55 +29,58 @@ if (!('telemetry' in window.MozillaResearch)) {
...
-    return function (command) {
-      if (navigator.doNotTrack === '1') { return; }
+    if (doNotTrack()) { return NO_OP; }
+
+    window._ga('create', trackingId, cookieDomain, name, fieldsObject);
+    name = typeof cookieDomain === 'object' ? cookieDomain.name : name;
+    return trackingFunction;
+    
+    function trackingFunction(command) {
+      if (doNotTrack()) { return; }
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I completely agree. I'll consider if doing here or in a follow-up.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 5]
```diff
@@ -13,55 +29,58 @@ if (!('telemetry' in window.MozillaResearch)) {
...
 };
 
 setupAnalytics();
 
 function setupAnalytics() {
-  window.ga = window.ga || function () {
-    (window.ga.q = (window.ga.q || [])).push(arguments)
+  if (doNotTrack()) { return; }
+
+  window.GoogleAnalyticsObject = '_ga';
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: Trying to explain this to you yesterday, I realize there is a simpler method for allowing the user to integrate with us requiring a minimal deviation from the standard process. I'll change the name of the PR and the documentation to reflect it.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 6]
```diff
@@ -13,64 +29,49 @@ if (!('telemetry' in window.MozillaResearch)) {
...
-        args[0] = command;
-      }
-      window.ga.apply(undefined, args);
-    };
-  }
-};
-
 var NO_OP = function () {};
 
+window.ga = NO_OP;
```
[the remark this answers is not in the document]
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: this breaks data collection for me:
screen shot 2018-03-19 at 9 24 37 pm
screen shot 2018-03-19 at 9 24 55 pm

if I remove this line, the commands seem to work like before:

screen shot 2018-03-19 at 9 22 59 pm
screen shot 2018-03-19 at 9 26 36 pm

<hr>

see my comments below for `setupAnalytics`

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 7]
```diff
@@ -13,64 +29,49 @@ if (!('telemetry' in window.MozillaResearch)) {
...
     }
   }
 };
 
 setupAnalytics();
 
 function setupAnalytics() {
+  if (doNotTrack()) { return; }
+
   window.ga = window.ga || function () {
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: can you change to this?

```js
var gaAlias = window.GoogleAnalyticsObject || 'ga';
window[gaAlias] = window[gaAlias] || function () {
  (window[gaAlias].q = window[gaAlias].q || []).push(arguments);
};
window.ga.l = +(new Date());
```

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: instead of `window.ga`, can you just create a property called like `window.MozillaResearch.telemetry._ga`? and use that elsewhere instead of `window.ga` (can of course create an intermediate variable like you do for `var telemetry`). that sound good?

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 8]
```diff
@@ -92,31 +93,51 @@ function startErrorLogging() {
...
-      if (navigator.doNotTrack === '1') {
-        return;
-      }
+      if (doNotTrack()) { return; }
+
       performance.measure(name, start, end);
       var performanceEntry = performance.getEntriesByName(name)[0];
       var duration = performanceEntry.duration;
-      ga('send', 'event', 'Performance', name, undefined, duration);
+      tracker('send', 'event', 'Performance', name, undefined, duration);
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I didn't know until I started work on the [hosted `telemetry.js` POC](https://github.com/MozillaReality/metrics/blob/5a01a9d/src/lib/reporter.js#L78-L86) (issue #193) that GA has an [API for perf timings](https://developers.google.com/analytics/devguides/collection/analyticsjs/user-timings).

so here instead of using a custom event by calling [`ga('send', 'event', 'Performance', …)`](https://developers.google.com/analytics/devguides/collection/analyticsjs/events), can we call [`ga('send', 'timing, …)`](https://developers.google.com/analytics/devguides/collection/analyticsjs/user-timings)?

```js
tracker('send', 'timing', name, 'load', duration);
```

<hr>

in case you want to handle this in another PR (or I can open one), I filed issue #208 to handle this. thanks!

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: btw, I got that [verbose logging](https://developers.google.com/analytics/devguides/collection/analyticsjs/debugging) from changing the GA script src to [**`https://www.google-analytics.com/analytics_debug.js`**](https://www.google-analytics.com/analytics_debug.js) – it's fantastic!

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 9]
```diff
@@ -92,31 +93,51 @@ function startErrorLogging() {
 
 function startAnalytics() {
   var CURRENT_VERSION = '1.0.1';
-  var ga = telemetry.ga.create('UA-77033033-6', 'auto', 'mozillaResearch');
-  ga('set', 'dimension1', CURRENT_VERSION);
-  ga('send', 'pageview');
-  return ga;
+  var tracker = newTracker('UA-77033033-6', 'auto', '__MozillaResearch');
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: To make the route very private. ;)

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 10]
```diff
@@ -92,31 +93,51 @@ function startErrorLogging() {
 
 function startAnalytics() {
   var CURRENT_VERSION = '1.0.1';
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: When we bumped the version, I'll change that. Moving to the top is a good idea.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 11]
```diff
@@ -92,31 +93,51 @@ function startErrorLogging() {
...
+
+  window.ga('create', trackingId, cookieDomain, name, fieldsObject);
+  name = typeof cookieDomain === 'object' ? cookieDomain.name : name;
+  return trackingFunction;
+
+  function trackingFunction(command) {
+    if (doNotTrack()) { return; }
+
+    var args = Array.prototype.slice.call(arguments);
+    if (command.indexOf('.') < 0 && name && command !== 'provide') {
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: It does not accept routing.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 12]
```diff
@@ -92,31 +93,51 @@ function startErrorLogging() {
...
       var performanceEntry = performance.getEntriesByName(name)[0];
       var duration = performanceEntry.duration;
-      ga('send', 'event', 'Performance', name, undefined, duration);
+      tracker('send', 'event', 'Performance', name, undefined, duration);
     }
   };
 }
 
-function injectScript (src, callback) {
+function newTracker(trackingId, cookieDomain, name, fieldsObject) {
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I'll comment it has the same signature of `create` command and provide a link.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 13]
```diff
@@ -2,6 +2,22 @@
 (function (window) {
 'use strict';
 
+/*
+ * The main requirements when dealing with Do-Not-Track (DNT) are:
+ *
+ *  1. Not to break the API.
+ *  2. Prevent explicit collection, storage and sending data if DNT is enabled.
+ *  3. Only sessions where DNT is disabled for its entire duration are expected to send
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: New visits or fresh reloads. I'll clarify.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 14]
```diff
@@ -2,6 +2,22 @@
 (function (window) {
 'use strict';
 
+/*
+ * The main requirements when dealing with Do-Not-Track (DNT) are:
+ *
+ *  1. Not to break the API.
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: Or GA, right. Since do not track needs to effectively disable all data sending and DNT is something not under the control of the developer, for a dev perspective, the API can not throw or break if DNT is enabled, it would simply do not work.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 8 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: Sure! It is better if we use a homogeneous API. Open a PR for that, please.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 15]
```diff
@@ -13,64 +29,49 @@ if (!('telemetry' in window.MozillaResearch)) {
...
     }
-  });  
+  });
 }
 
-function startErrorLogging() {
+function setupErrorLogging() {
+  if (doNotTrack()) { return; }
+
   injectScript('https://cdn.ravenjs.com/3.22.3/console/raven.min.js', function (err) {
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: Perfect, the fewer different APIs we use, the better. Feel free to implement that.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 16]
```diff
@@ -2,6 +2,22 @@
...
 
+/*
+ * The main requirements when dealing with Do-Not-Track (DNT) are:
+ *
+ *  1. Not to break the API.
+ *  2. Prevent explicit collection, storage and sending data if DNT is enabled.
+ *  3. Only sessions where DNT is disabled for its entire duration are expected to send
+ *     telemetry data.*
+ *
+ * Some collection can happen implicitly when loading 3rd party libraries. It is
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: Notice this is a note describing design decisions we do in the telemetry API, here I'm explaining we can not track all the scripts we are using but we can do our best effort to avoid this issue.

[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I'll clarify anyway.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 17]
```diff
@@ -13,64 +29,49 @@ if (!('telemetry' in window.MozillaResearch)) {
...
 
-  if (navigator.doNotTrack === '1') {
-    return;
-  }
   injectScript('https://www.google-analytics.com/analytics.js', function (err) {
     if (err) {
       console.warn('Could not load Analytics.js script:', err);
       return;
     }
-  });
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: That's because GA is always created (unless DNT is on). It is not created inside the script, it is only consumed by it. I was not able to find the un-minified `analytics.js` script so I don't know what should I check to effectively know if the script has loaded.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 18]
```diff
@@ -13,64 +32,57 @@ if (!('telemetry' in window.MozillaResearch)) {
...
 
-  if (navigator.doNotTrack === '1') {
-    return;
-  }
-  injectScript('https://www.google-analytics.com/analytics.js', function (err) {
+  window.dataLayer = window.dataLayer || [];
+  window.gtag = window.gtag || function () { dataLayer.push(arguments); };
+  window.gtag('js', new Date());
+
+  injectScript('https://www.googletagmanager.com/gtag/js?id=UA-77033033-6', function (err) {
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: can you put the ID in a constant at the top of the file?

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 19]
```diff
@@ -80,43 +92,72 @@ function startErrorLogging() {
...
-    configureRaven();
+    startRaven();
   });
 
-  function configureRaven () {
-    console.log('Raven.js script loaded');
+  function startRaven () {
     Raven.config('https://e359be9fb9324addb0dc97b664cf5ee6@sentry.io/294878')
-         .install();
+          .install();
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: was this space intentional? 6 spaces? I'd think either 2 spaces, 4 spaces, or aligned with the `.` in the line above

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 20]
```diff
@@ -80,43 +92,72 @@ function startErrorLogging() {
...
   }
 };
 
 function startAnalytics() {
-  var CURRENT_VERSION = '1.0.1';
-  var ga = telemetry.ga.create('UA-77033033-6', 'auto', 'mozillaResearch');
-  ga('set', 'dimension1', CURRENT_VERSION);
-  ga('send', 'pageview');
-  return ga;
+  var tracker = newTracker('UA-77033033-6', {
```
[the remark this answers is not in the document]
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: is supplying the ID actually necessary now with `gtag.js`? this is defined in the [`<script src="…gtag.js…?id=…">` above](https://github.com/mozilla/unity-webvr-export/pull/200/files#diff-f080f296c7d77ffcb43fcbe4160e91a3R75). if it's unnecessary, can we change the function argument signature of `newTracker` to not require it?

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 21]
```diff
@@ -80,43 +92,72 @@ function startErrorLogging() {
...
+    'groups': 'MozillaResearch',
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I feel like this should be passed as an argument, but I wouldn't say it blocks merging of this

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 22]
```diff
@@ -80,43 +92,72 @@ function startErrorLogging() {
...
+    'custom_map': {
+      'dimension1': 'version'
+    }
+  });
+  tracker('event', 'using_webvr_template', { version: CURRENT_VERSION });
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: could we name this `unity_assets_version` instead?

Comment (cvan) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: is this easy to filter in GA aggregations? I have experience with GA, but I don't do any fancy queries, so I'm not sure if a string vs. object is easier to parse and filter by. do you know off hand?

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 23]
```diff
@@ -134,4 +175,13 @@ function injectScript (src, callback) {
   return script;
 }
 
-})(window);
+// IE9/IE10 uses a prefixed version while MS Edge sets the property in
+// `window` instead of `navigator`:
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: perfect - thank you!

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 24]
```diff
@@ -134,4 +175,13 @@ function injectScript (src, callback) {
...
+// https://developer.mozilla.org/en-US/docs/Web/API/Navigator/doNotTrack#Browser_compatibility
+function doNotTrack() {
+  return navigator.doNotTrack === '1' ||
+          navigator.msDoNotTrack === '1' ||
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: is the indentation intentional here?

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 18 above)
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: btw, I've been meaning to ask: how do we test that GA is collecting the metrics correctly? I know with [`analytics.js`](https://developers.google.com/analytics/devguides/collection/analyticsjs/debugging), you can change the source to `analytics_debug.js` to get verbose logging for development/testing.

besides checking that the GA calls are formed correctly and the XHRs/`navigator.sendBeacon` calls get correctly posted with acceptable HTTP responses, is there a way to check that the data is submitted and _stored_ correctly? in my tests, with my private GA tracking ID, I don't see any data stored. did you have to flip any prefs for this particular GA property to get cross-origin tracking to work?

in other words, if I'm going to test this script with my GA ID, which config settings do I need to enable to verify the telemetry calls are properly being stored by GA?

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: It is not needed to get the code but GA keeps track of who initiated the download of the script for an specific site.

[the remark this answers is not in the document]
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: gotcha, understood - I'm not very familiar with `gtag.js`. I'd say just put the GA ID in a variable at the top and reuse it in the two spots. a comment would be useful, but beyond my curiosity, I'm not sure it's really useful. thanks for answering

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 18 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I was not able to find anything similar but I did find a [Chrome extension provided by the GA team](https://chrome.google.com/webstore/detail/google-analytics-debugger/jnkmfdileelhofjcijamephohjechhna). It seems `gtag` is using `analytics.js` or something similar under the hood because it is using the very same API but `gtag` requires less code.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 21 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I thought about it and I agree. Since this API is used internally only right now, I don't mind to change it but there were some edge cases I wanted to avoid so I chose this approach. I'll give it a spin.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 22 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I prefer to keep the name. It seems to me that `gtag` managed to convert custom dimensions and other events to _parameters_ for events so it makes sense to say "using the WebVR template at version X.Y.X". But I'll change if GA ergonomics are bad. Let's improve on the top of something.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 25]
```diff
@@ -80,43 +92,72 @@ function startErrorLogging() {
...
+  var groups = options.groups;
+  telemetry._gtag('config', trackingId, options);
+  return trackingFunction;
+
+  function trackingFunction(command, label, options) {
+    if (doNotTrack()) { return; }
+
+    options = options || {};
+    if (groups) {
+      options.send_to = groups;
```
[the remark this answers is not in the document]
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: The `newTracker` method returns a "bound-to-groups" tracking function. It is for internal use only, it is a helper indeed and it's not exposed as part of the public API. I'm going to clarify in the explanation though.

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 26]
```diff
@@ -13,64 +33,52 @@ if (!('telemetry' in window.MozillaResearch)) {
...
 
-  if (navigator.doNotTrack === '1') {
-    return;
-  }
-  injectScript('https://www.google-analytics.com/analytics.js', function (err) {
+  window.dataLayer = window.dataLayer || [];
+  window.gtag = window.gtag || function () { dataLayer.push(arguments); };
+  window.gtag('js', new Date());
+
+  injectScript('https://www.googletagmanager.com/gtag/js?id=' + MOZILLA_RESEARCH_TRACKER, function (err) {
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: perfect - thanks

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 27]
```diff
@@ -80,43 +88,83 @@ function startErrorLogging() {
...
+/**
+ * The function configures Google Analytics sending a `config` command [1] but
+ * also returns a function to use instead of `gtag` that respects `Do-Not-Track`
+ * and it's bound to the `groups` property of the `options` parameter to avoid
+ * adding the `send_to` key [2] to each command.
+ *
+ * [1] https://developers.google.com/analytics/devguides/collection/gtagjs/pages
+ * [2] https://developers.google.com/analytics/devguides/collection/gtagjs/sending-data#groups-and-properties
+ *
+ * @param {string} trackingId see `'config'` command signature [1]
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: awesome! thank you for going above and beyond with the JSDoc documentation here :+1:

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 28]
```diff
@@ -80,43 +88,83 @@ function startErrorLogging() {
...
+ * @param {onject} options see `'config'` command signature [1]
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: typo: `onject` -> `object`

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 29]
```diff
@@ -80,43 +88,83 @@ function startErrorLogging() {
...
+ * @returns {function} A tracker function to replace invocation of `gtag` that
+ * honours Do-Not-Track and automatically adds the `send_to` key to the
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: I approve of the British spelling, `honours` 👍

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (cvan) [reply Assets/WebGLTemplates/WebVR/lib/telemetry.js]: ah, gotcha - this makes sense. thanks for the explanation and comment. much appreciated

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 30]
```diff
@@ -134,4 +182,13 @@ function injectScript (src, callback) {
   return script;
 }
 
+// IE9/IE10 uses a prefixed version while MS Edge sets the property in
+// `window` instead of `navigator`:
+// https://developer.mozilla.org/en-US/docs/Web/API/Navigator/doNotTrack#Browser_compatibility
+function doNotTrack() {
+  return navigator.doNotTrack === '1' ||
+         navigator.msDoNotTrack === '1' ||
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: thanks for fixing the nit indentation

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 31]
```diff
@@ -13,64 +33,52 @@ if (!('telemetry' in window.MozillaResearch)) {
...
+    return window.gtag || NO_OP;
+  }
+});
+
 telemetry.performance = {
   mark: NO_OP,
   measure: NO_OP
 };
 
 telemetry.start = function (config) {
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: not sure why I didn't think of it before, but we should probably have an `if` statement inside here that ensures that this method wasn't already called. (in other words, calling `MozillaResearch.telemetry.start()` multiple times shouldn't have unintended side effects.)
